### PR TITLE
server: reworking factory contexts

### PR DESCRIPTION
**Pull request by alyssawilk**
This is a large and toil-y refactoring of FactoryContexts in preparation for upstream filters and ECDS. We went through several design options  (e.g. https://github.com/envoyproxy/envoy/pull/22061) and ended up with this, which is not trivial to update to, but provides a cleaner long term structure to factories and ownership.

We apologize for the downstream burden for folks pulling this PR.  In most cases you can
1) change createFilterFactoryFromProtoTyped to have Server::Configuration::FactoryContext& context
2) add a local member variable Server::Configuration::ServerFactoryContext& context = base_context.getServerFactoryContext();
3) update any calls to initManager or scope to use base_context.

(3) is especially important as it ensures that things work as expected for LDS, using the downstream init manager, and downstream stats scope.  It also makes this a very high risk PR - anything which will not persist beyond the downstream context must use the server context and the prior mixed inheritance model made it very unclear which was used where and when.

Risk Level: VERY HIGH
Testing: all the tests _finally_ pass
Docs Changes: n/a
Release Notes: 
Part of https://github.com/envoyproxy/envoy/issues/10455

### CODE REVIEW

File: envoy/server/factory_context.h [hunk 1]
```diff
@@ -264,12 +269,44 @@ class FactoryContext : public virtual CommonFactoryContext {
...
+  FilterFactoryContext(ServerFactoryContext& server_context,
+                       OptRef<DownstreamFactoryContext> downstream_context)
+      : server_context_(server_context), downstream_context_(downstream_context) {}
+
+  /**
+   * @return ServerFactoryContext which lifetime is no shorter than the server.
+   */
+  ServerFactoryContext& getServerFactoryContext() { return server_context_; }
+
+  OptRef<DownstreamFactoryContext> getDownstreamFactoryContext() { return downstream_context_; }
```
Comment (alyssawilk) [inline envoy/server/factory_context.h:286]: @kyessenov downstreamFactoryContext would be available for downstream filters, not for upstream filters.  Lifetime would be lifetime of the downstream listener

File: envoy/server/factory_context.h [hunk 2]
```diff
@@ -264,12 +269,44 @@ class FactoryContext : public virtual CommonFactoryContext {
   virtual ProcessContextOptRef processContext() PURE;
 };
 
+/**
+ * Context for downstream network and HTTP filters.
+ */
```
Comment (mattklein123) [inline envoy/server/factory_context.h:274]: Is this comment out of date or if it is actually the context for downstream network and http filters why does getDownstreamFactoryContext() return an OptRef? Shouldn't it always have that context?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks I'm a huge +1 on this approach. IMO it is vastly simpler to understand. Thank you for iterating on this. @kyessenov wdyt?

**Comment by alyssawilk** [comment]
Ok having spent 2 more days swinging at this, I want to check in again.
I agree the abstractions are much more clean, but I will say that the API changes for folks implementing filters are super toily.
I managed to minimize the toil for HTTP filters, by changing the factories such that by default we use the ServerFactoryContext and don't have to getServerFactoryContext() left and right, but ironically because of the way network factories are already split out I couldn't do it there.

I've got test/integration and all extensions building, but just want to check in that we're OK with a more painful API change for the various features we're getting out of this in the long run before I go tackle tests again.

File: envoy/server/factory_context.h [hunk 3]
```diff
@@ -246,56 +257,38 @@ class DownstreamFactoryContext {
...
     if (getDownstreamFactoryContext().has_value()) {
       return getDownstreamFactoryContext()->messageValidationVisitor();
     }
     return getServerFactoryContext().messageValidationVisitor();
   }
 
-private:
-  ServerFactoryContext& server_context_;
-  OptRef<DownstreamFactoryContext> downstream_context_;
+  operator ServerFactoryContext&() { return this->getServerFactoryContext(); }
```
Comment (mattklein123) [inline envoy/server/factory_context.h:291]: I get why you did this, but pretty magical and I'm not sure we really want this?

**Review by mattklein123 — COMMENTED** [review]
Thanks for slogging through this. I'm +1 on this approach even if it causes some build/integration pain for people. I think we can have some good release notes which should make it pretty clear what to do. I think this is going to make things much easier to understand and extend later, and also make the upstream stuff and potential sharing much cleaner. Thank you!

**Comment by alyssawilk** [comment]
Ok, one full week plus some overtime and this is good to go.  All tests pass but the sip proxy test I filed https://github.com/envoyproxy/envoy/issues/22306 for.

Also please see updated PR description.  This PR now scares the heck out of me.  Tests caught 2 lifetime issues and I'm increasingly concerned about hard to find stats leak due to misplaced scope()

**Comment by jmarantz** [comment]
I haven't looked at the 299 files yet :). Scared :)

Can you add some description to the PR about what motivated this change? Is it strictly https://github.com/envoyproxy/envoy/issues/10455 ?

**Comment by alyssawilk** [comment]
@jmarantz "This is a large and toil-y refactoring of FactoryContexts in preparation for upstream filters and ECDS." not clear enough?
Originally this was so we could reuse shared code for upstream HTTP filters but it fixes some issues Kuat had with ECDS.  If it were just upstream filters I think I would have gone the template route by now.

**Comment by RyanTheOptimist** [comment]
Any idea what's up with the failing tests? Are they flakes, or are they related to the changes in this PR?

**Comment by alyssawilk** [comment]
/wait on CI
I think they're largely CI being overloaded because nothing is cached, but I want to land https://github.com/envoyproxy/envoy/pull/22301 and see if I can get away without this change.

**Comment by alyssawilk** [comment]
OK, having thought about this some more (and trying to avoid the sunk cost fallacy) I'm REALLY leery of landing this PR.  
The lifetime issues I hit when debugging the final tests mean that not only is it super toily for users, it's also really high risk. 
Soooo, I went and tried another tack, and spent a day rebasing my upstream filters prototype (which now passes some of the inteegration tests!) with a template based solution
full PR complete with way too many FIXMES here:
https://github.com/envoyproxy/envoy/compare/main...alyssawilk:upstream_filters_clean?expand=1
the template changes are largely here:
https://github.com/envoyproxy/envoy/commit/53d466fac48e1697584b86addda678cca9114ff9

It's a lot uglier (god help whoever has to touch the config discovery templates next) but from buffer/config.* you can see the work to take a downstream factory and make it an upstream factory is really minimal, and having an opt-in system isn't that bad.

So unless @kyessenov or @mattklein123 actively want the filter refactor, I'm inclined to abandon this PR in favor of the template scariness (which is far less toil and risk.)

**Comment by alyssawilk** [comment]
see https://github.com/envoyproxy/envoy/pull/22358 for a cleaner diff

//test/integration:integration test passes, but there might be a few other unit test fix ups I need to make for templates

**Pull request closed without merging**